### PR TITLE
Unify element activation, element death and moving interface

### LINKED ISSUE #17100: Unify element activation, element death and moving interface
## Reason

Currently we have ActivateElementsUserObject to deal with element activation. But we could really reuse the code for element death and moving interface as well.

## Design

Rename the class and generalize. (Along with several parallel bug fixes.)

## Impact

New API.

### PR DESCRIPTION
**Pull request by hugary1995**
Generalize the existing ActivateElementUserObjects, add ElementSubdomainModifiers, and fix some parallel bugs.

close #17100

### CODE REVIEW

**Comment by hugary1995** [comment]
@fdkong @dschwen Could you please review this since you reviewed the original #16008 PR about element activation?

The new UO changes some of the logic (and names of several parameters, variables, etc) in the original UO so that element activation can be generalized to element death and moving interface.

The original regression tests didn't capture several parallel nodeset/sideset generation issues. I fixed those and added more regression tests.

The [documentation](https://mooseframework.inl.gov/docs/PRs/17101/site/source/userobject/CoupledVarThresholdElementSubdomainModifier.html) should be helpful. It also explicitly mentions what we can do with this UO.

This PR will allow me to couple the element subdomain modifier with XFEM, to get smooth boundary/interface instead of the zig-zag boudaries. @bwspenc the XFEM element subdomain modifier will have to wait for this one to be merged first.

File: framework/src/userobject/ElementSubdomainModifier.C
Comment (hugary1995) [inline framework/src/userobject/ElementSubdomainModifier.C:138]: I don't know if I need to also project initial conditions on displaced elements. I am only projecting ICs on undisplaced elements.

File: test/tests/userobjects/element_subdomain_modifier/block_restricted.i
Comment (hugary1995) [inline test/tests/userobjects/element_subdomain_modifier/block_restricted.i:37]: There is still a bug I need to fix: if I don't add INITIAL here, the "moving_boundary" nodeset/sideset won't be created.

**Review by dewenyushu — DISMISSED** [review]
Looks great!

This PR is a better and generalized version of `ActivateElementsUserObjectBase`. Maybe we can later convert the tests from the old element activation code and combine these two.

File: test/tests/userobjects/element_subdomain_modifier/block_restricted.i
Comment (hugary1995) [reply test/tests/userobjects/element_subdomain_modifier/block_restricted.i:37]: Update: this is due to the current limitation that libmesh doesn't output empty nodeset/sideset. This will be fixed after [https://github.com/libMesh/libmesh/pull/2871](https://github.com/libMesh/libmesh/pull/2871).

Comment (loganharbour) [reply test/tests/userobjects/element_subdomain_modifier/block_restricted.i:37]: Do you plan on waiting for a libmesh update to get this in or do you want to get it in first? If the latter - make an issue and be sure to reference the issue on the test.

Comment (hugary1995) [reply test/tests/userobjects/element_subdomain_modifier/block_restricted.i:37]: There is no time pressure on this, so I'm fine either way. I prefer to get this in first (and a regression test for the empty nodeset/sideset after a libmesh update). But I'm open to the other option if you think it is better.

Comment (hugary1995) [reply test/tests/userobjects/element_subdomain_modifier/block_restricted.i:37]: There is also a possibility that a libmesh update comes before this PR is ready to merge.

Comment (loganharbour) [reply test/tests/userobjects/element_subdomain_modifier/block_restricted.i:37]: I have no preference. Just if you choose to get it in before the libmesh update, please create an issue and reference it in the test/test spec

File: framework/include/userobject/ElementSubdomainModifier.h
Comment (fdkong) [inline framework/include/userobject/ElementSubdomainModifier.h:48]: I am curious about if we are able to de-reference the object directly from the smart pointer (instead of raw pointer)?

File: framework/include/userobject/ElementSubdomainModifier.h
Comment (fdkong) [inline framework/include/userobject/ElementSubdomainModifier.h:53]: Do you mind adding comments for all of the following functions?

File: framework/src/userobject/ElementSubdomainModifier.C
Comment (fdkong) [inline framework/src/userobject/ElementSubdomainModifier.C:18]: We may not need this header if it is already a parallel object

File: framework/src/userobject/ElementSubdomainModifier.C
Comment (fdkong) [inline framework/src/userobject/ElementSubdomainModifier.C:63]: double `{` ?

File: framework/src/userobject/ElementSubdomainModifier.C
Comment (fdkong) [inline framework/src/userobject/ElementSubdomainModifier.C:64]: `mooseAssert(boundary_ids.size==1, ".... ... ")`

File: framework/src/userobject/ElementSubdomainModifier.C
Comment (fdkong) [inline framework/src/userobject/ElementSubdomainModifier.C:91]: `elem` will be different from `_current_elem`?  That being said, can we directly use `_current_elem` instead of `elem` in the following several places?

File: framework/src/userobject/ElementSubdomainModifier.C
Comment (fdkong) [inline framework/src/userobject/ElementSubdomainModifier.C:153]: I feel like it is better to  change "`moved`" to `activated`.  `activated` makes more sense for in this context

File: framework/src/userobject/ElementSubdomainModifier.C
Comment (fdkong) [inline framework/src/userobject/ElementSubdomainModifier.C:330]: `ElementSubdomainModifier` has to use with a transient executioner?  If not, we need to check if we need to use old and older solutions

File: framework/src/userobject/ElementSubdomainModifier.C
Comment (fdkong) [inline framework/src/userobject/ElementSubdomainModifier.C:350]: We do not need `close` for current_solution since we do not really change values

File: framework/src/userobject/ElementSubdomainModifier.C
Comment (fdkong) [inline framework/src/userobject/ElementSubdomainModifier.C:347]: Why `older_solution` does not take solution from `old_solution` ?

File: framework/src/userobject/ThresholdElementSubdomainModifier.C
Comment (fdkong) [inline framework/src/userobject/ThresholdElementSubdomainModifier.C:64]: `ThresholdElementSubdomainModifier` will handle only two subdomains?

**Review by fdkong — CHANGES_REQUESTED** [review]
Looks very good for me.

File: framework/include/userobject/ElementSubdomainModifier.h
Comment (loganharbour) [reply framework/include/userobject/ElementSubdomainModifier.h:48]: You can remove the `.get()`. The dereference operator for a smart pointer returns a reference to the underlying object.

File: framework/src/userobject/ElementSubdomainModifier.C
Comment (hugary1995) [reply framework/src/userobject/ElementSubdomainModifier.C:144]: I haven't seen any issues so far without initializing solutions on the displaced mesh. But as you said there maybe corner cases I'm not aware of. I'll just rely on `syncSolutions()` for now.

File: test/tests/userobjects/element_subdomain_modifier/block_restricted.i
Comment (hugary1995) [reply test/tests/userobjects/element_subdomain_modifier/block_restricted.i:37]: Okay I created an issue and referenced it.

File: framework/include/userobject/ElementSubdomainModifier.h
Comment (hugary1995) [reply framework/include/userobject/ElementSubdomainModifier.h:48]: You are right, I removed the `.get()`s.

File: framework/include/userobject/ElementSubdomainModifier.h
Comment (hugary1995) [reply framework/include/userobject/ElementSubdomainModifier.h:53]: Added.

File: framework/src/userobject/ElementSubdomainModifier.C
Comment (hugary1995) [reply framework/src/userobject/ElementSubdomainModifier.C:18]: Removed.

File: framework/src/userobject/ElementSubdomainModifier.C
Comment (hugary1995) [reply framework/src/userobject/ElementSubdomainModifier.C:63]: I think I got min gcc compiler error if I don't use double `{`.

File: framework/src/userobject/ElementSubdomainModifier.C
Comment (hugary1995) [reply framework/src/userobject/ElementSubdomainModifier.C:64]: Added.

File: framework/src/userobject/ElementSubdomainModifier.C
Comment (hugary1995) [reply framework/src/userobject/ElementSubdomainModifier.C:91]: But `_current_elem` is `const`, so I can't use it to directly change the subdomain ID.

File: framework/src/userobject/ElementSubdomainModifier.C
Comment (hugary1995) [reply framework/src/userobject/ElementSubdomainModifier.C:153]: I don't think so. The elements here really just changed subdomain IDs. `activated` would make perfect sense if we use this object for element activation, but would be misleading in the context of element deletion and moving interface.

File: framework/src/userobject/ElementSubdomainModifier.C
Comment (hugary1995) [reply framework/src/userobject/ElementSubdomainModifier.C:330]: How do I check that?

File: framework/src/userobject/ElementSubdomainModifier.C
Comment (hugary1995) [reply framework/src/userobject/ElementSubdomainModifier.C:347]: So I am only setting ICs when `apply_ic = true`. That is meant for element activation/deletion, which implies that these dofs just moved from an inactive domain to an active domain, or vice versa. Therefore, I am applying the IC as their current/old/older solutions -- mimicking an actual IC.

Comment (hugary1995) [reply framework/src/userobject/ElementSubdomainModifier.C:347]: Let me know whether you think this makes sense or not.

File: framework/src/userobject/ThresholdElementSubdomainModifier.C
Comment (hugary1995) [reply framework/src/userobject/ThresholdElementSubdomainModifier.C:64]: Right now yes. Actually if the user needs more than one thresholds (more than two subdomains to switch between), they can add multiple `ThresholdElementSubdomainModifer`s in their input file.

File: framework/src/userobject/ElementSubdomainModifier.C
Comment (hugary1995) [reply framework/src/userobject/ElementSubdomainModifier.C]: Right.

File: framework/src/userobject/ElementSubdomainModifier.C
Comment (hugary1995) [reply framework/src/userobject/ElementSubdomainModifier.C:144]: Hold on, I actually got segfault if I call `_displaced_problem->syncSolutions()`. Not sure why.

**Comment by dschwen** [comment]
What's the deal with the empty jpg files added in this PR?

**Comment by hugary1995** [comment]
> What's the deal with the empty jpg files added in this PR?

I don't know. The images are actually not empty if you check them out. I guess it's a github issue?

**Comment by hugary1995** [comment]
@fdkong could you take another look at it? When I said "there is no time pressure on this" I wasn't expecting this to take two weeks. I have a report due in April. Thanks.

File: framework/src/userobject/ElementSubdomainModifier.C
Comment (fdkong) [reply framework/src/userobject/ElementSubdomainModifier.C:347]: This makes sense if it is applied to only "new" elements

File: framework/src/userobject/ElementSubdomainModifier.C
Comment (fdkong) [reply framework/src/userobject/ElementSubdomainModifier.C:330]: You may always need a transient? The element activation won't happen for a steady simulation. 

If this assumption is right, we could merge the code as is

**Comment by hugary1995** [comment]
@fdkong I followed your suggestions and added a regression test for the steady case.

File: framework/src/userobject/ElementSubdomainModifier.C
Comment (hugary1995) [reply framework/src/userobject/ElementSubdomainModifier.C]: This is supposed to work in steady as well. I fixed it.

File: framework/src/userobject/CoupledVarThresholdElementSubdomainModifier.C
Comment (dschwen) [inline framework/src/userobject/CoupledVarThresholdElementSubdomainModifier.C:37]: should this be multiplied by `_JxW[_qp] * _coord[_qp]`?

Comment (hugary1995) [reply framework/src/userobject/CoupledVarThresholdElementSubdomainModifier.C]: Good catch! I fixed that. @dewenyushu this will probably improve some of your AM results.

**Review by dschwen — APPROVED** [review]

File: framework/src/userobject/CoupledVarThresholdElementSubdomainModifier.C
Comment (fdkong) [reply framework/src/userobject/CoupledVarThresholdElementSubdomainModifier.C]: It is indeed a good catch!!

File: framework/src/userobject/ElementSubdomainModifier.C
Comment (fdkong) [inline framework/src/userobject/ElementSubdomainModifier.C:330]: I like it!

**Review by fdkong — APPROVED** [review]
Looks great

File: framework/src/userobject/CoupledVarThresholdElementSubdomainModifier.C
Comment (dewenyushu) [reply framework/src/userobject/CoupledVarThresholdElementSubdomainModifier.C]: Oops I guess so! I had trouble with element activation using coupled variable. Maybe this is one of the reasons! Nice catch!

**Comment by loganharbour** [comment]
@hugary1995 this is failing rather painfully almost everywhere in parallel: https://civet.inl.gov/event/65489/

Are you able to look into this asap or do we need to?

**Comment by hugary1995** [comment]
Sure! I'll look at it ASAP!

**Comment by hugary1995** [comment]
@loganharbour I can't figure it out any time soon. Can we revert it?

**Pull request merged**
